// plxy-grid/packages/conf/index.js
const GlobalConfig = {
  // showOverflow: null,
  // showHeaderOverflow: null,
  // resizeInterval: 250,
  // size: null,
  // radioConfig: {
  //   trigger: 'default'
  // },
  // checkboxConfig: {
  //   trigger: 'default'
  // },
  // sortConfig: {
  //   remote: false,
  //   trigger: 'default'
  // },
  // filterConfig: {
  //   remote: false
  // },
  // expandConfig: {
  //   trigger: 'default'
  // },
  // treeConfig: {
  //   children: 'children',
  //   hasChild: 'hasChild',
  //   indent: 20
  // },
  // tooltipConfig: {
  //   theme: 'dark'
  // },
  // validConfig: {
  //   message: 'default'
  // },
  // editConfig: {
  //   mode: 'cell'
  // },
  // resizable: false,
  // stripe: false,
  // border: false,
  fit: true,
  emptyCell: '　',
  showHeader: true,
  zIndex: 100,
  rowId: '_XID', // 行数据的唯一主键字段名
  version: 0, // 版本号，对于某些带数据缓存的功能有用到，上升版本号可以用于重置数据
  optimization: {
    animat: true,
    delayHover: 250,
    // rHeights: {
    //   default: 48,
    //   medium: 44,
    //   small: 40,
    //   mini: 36
    // },
    scrollX: {
      gt: 60
      // oSize: 0,
      // rSize: 0
      // vSize: 0
    },
    scrollY: {
      gt: 100
      // oSize: 0,
      // rSize: 0
      // vSize: 0,
      // rHeight: 0
    }
  },
  icon: {
    sortAsc: 'plx-icon--caret-top',
    sortDesc: 'plx-icon--caret-bottom',
    filterNone: 'plx-icon--funnel',
    filterMatch: 'plx-icon--funnel',
    edit: 'plx-icon--edit-outline',
    treeLoaded: 'plx-icon--refresh roll',
    treeOpen: 'plx-icon--caret-right rotate90',
    treeClose: 'plx-icon--caret-right',
    expandLoaded: 'plx-icon--refresh roll',
    expandOpen: 'plx-icon--arrow-right rotate90',
    expandClose: 'plx-icon--arrow-right',
    refresh: 'plx-icon--refresh',
    refreshLoading: 'plx-icon--refresh roll',
    import: 'plx-icon--upload',
    importRemove: 'plx-icon--close',
    export: 'plx-icon--download',
    zoomIn: 'plx-icon--zoomin',
    zoomOut: 'plx-icon--zoomout',
    custom: 'plx-icon--menu',
    jumpPrev: 'plx-icon--d-arrow-left',
    jumpNext: 'plx-icon--d-arrow-right',
    prevPage: 'plx-icon--arrow-left',
    nextPage: 'plx-icon--arrow-right',
    jumpMore: 'plx-icon--more',
    modalZoomIn: 'plx-icon--zoomin',
    modalZoomOut: 'plx-icon--zoomout',
    modalClose: 'plx-icon--close',
    modalInfo: 'plx-icon--info',
    modalSuccess: 'plx-icon--success',
    modalWarning: 'plx-icon--warning',
    modalError: 'plx-icon--error',
    modalQuestion: 'plx-icon--question',
    modalLoading: 'plx-icon--refresh roll',
    caretBottom: 'plx-icon--caret-bottom',
    dropdownBottom: 'plx-icon--arrow-bottom',
    btnLoading: 'plx-icon--refresh roll'
  },
  grid: {
    proxyConfig: {
      autoLoad: true,
      message: true,
      props: {
        list: null,
        result: 'result',
        total: 'page.total'
      },
      beforeColumn: null,
      beforeQuery: null,
      beforeDelete: null,
      beforeSave: null
    }
  },
  menu: {},
  tooltip: {
    trigger: 'hover',
    theme: 'dark',
    leaveDelay: 300
  },
  pager: {
    // perfect: true,
    // pageSize: 10,
    // pagerCount: 7,
    // pageSizes: [10, 15, 20, 50, 100],
    // layouts: ['PrevJump', 'PrevPage', 'Jump', 'PageCount', 'NextPage', 'NextJump', 'Sizes', 'Total']
  },
  toolbar: {
    // import: {
    //   mode: 'covering'
    // },
    // export: {
    //   types: ['csv', 'html', 'xml', 'txt']
    // },
    // resizable: {
    //   storage: false
    // },
    // custom: {
    //   storage: false,
    //   isFooter: true
    // },
    // buttons: []
  },
  modal: {
    minWidth: 340,
    minHeight: 200,
    lockView: true,
    mask: true,
    duration: 3000,
    marginSize: 8,
    dblclickZoom: true,
    animat: true
  },
  i18n: key => key
}

export default GlobalConfig
